Adds a tool to count exon overlapping reads.

vruano opened the pull request:
Resolves functionality issue #344.

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,681 @@
+       fullName="countColumnName",shortName="countColumnName")
+    protected String countColumnName = "%^";
+
+    private PrintWriter outputWriter;
```
Sure.

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,681 @@
+        logger.log(Level.INFO,"Calculating exon depths...");
+    }
+
+    private List<GenomeLoc> fromSimpleIntervalsToGenomeLocs(final GenomeLocParser genomeLocFactory,
```
Thanks.

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,681 @@
+            return Collections.emptyList();
+        else if (intervals.size() == 1) {
+            final Locatable locatable = intervals.get(0);
+            return Collections.singletonList(genomeLocFactory.createGenomeLoc(locatable.getContig(),locatable.getStart(),locatable.getEnd()));
```
Since I'm moving this to a general utility class the comment above no longer applies.

akiezun replied:
premature optimization is always wrong. The only OK optimization is that backed by profile data.

droazen replied:
This view ("premature optimization is always wrong. The only OK optimization is that backed by profile data.") is a bit too extreme, I think. Even before profiling, we should not be intentionally doing egregiously inefficient things. For example, it's good practice to use `StringBuilder` when incrementally building up a large String instead of string concatenation. It's only when the optimization falls into the category of a micro-optimization that seriously harms code readability that we should start screaming "premature optimization".

akiezun replied:
using `StringBuilder` matters only if it's in a tight loop. Otherwise it does not matter. 

In this case, my complaint was exactly about readability - having a special branch for the case of 1 is premature optimization.

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,681 @@
+        writeExonOutputHeaders(outputWriter, summaryPerRowOutputWriter);
+        final int exonCount = exonDB.exonCount();
+        final int countColumnCount = counts.columnCount();
+        final int[] countBuffer = new int[countColumnCount];
```
Repeatingly creating one element int array hurts me.

akiezun replied:
premature optimization. The human reader of this code is the most important consideration. Performance tuning will come later and we'll use a profiler.

vruano replied:
I am done some refactoring that avoid copy all together, so no problem there any longer.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,681 @@
+
+    @Override
+    public void onTraversalStart() {
+        super.onTraversalStart();
```
fine. but unused code is just overhead for the reader because now i have to read the super method and then i find out it's empty so I lost time and context.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,681 @@
+    }
+
+    private Counts createCounts() {
+        switch(groupBy) {
```
the fact that there this enum exists and it has the same items as the 'Counts' hierarchy is a bad code smell. It's duplicating the logic of saying that there are 3 ways of grouping coverage counts.

vruano replied:
mmm will see what I can do...

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,681 @@
+
+        // A rough generous estimate of 256 bytes per count column + 1024 for the header.
+        final StringBuilder sb = new StringBuilder(counts.columnCount() * 256 + 1024);
+        sb.append(COMMENT_PREFIX + "fileFormat = tsv").append('\n');
```
premature optimization. Keep the human reader of the code in mind - optimize for him/her not for the machine.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,681 @@
+        }
+
+        @Override
+        public long[] columnSumsToArray(final long[] dest) {
```
`Collection.toArray` has no other choice because of deficient type system in Java (you can't call `new T[size]` so they need to take in an array). The consideration here is completely different. And the harm is very real - the human reader is tricked by the unexpected side effect of this call.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,681 @@
+        }
+
+        @Override
+        public int columnCount() {
```
premature optimization

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,681 @@
+
+        private final int sampleCount;
+
+        private boolean foundMissingSample = false;
```
yes, just kill quickly

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExonDB.java:
```diff
@@ -0,0 +1,370 @@
+    }
+
+    /**
+     * Returns the exome-size in base-pairs as the sum of all its exon's sizes.
```
`Returns the size of the exome in base-pairs (computed as the sum of all its exons sizes).`

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonDB.java:
```diff
@@ -0,0 +1,361 @@
+        sortedIntervals = new ArrayList<>();
+        if (intervals.isEmpty())
+            return;
+        int nextIndex = 0;
```
can you extract this code below to a new method? Constructors should be as simple as possible and this code does some work - at least by extracting a method it will have a clearer purpose

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonDB.java:
```diff
@@ -0,0 +1,361 @@
+     *
+     * @return can return {@code null} indicating that this interval should remain anonymous.
+     */
+    protected String intervalName(@SuppressWarnings("unused") final int index, final GenomeLoc location) {
```
if you don't use the variable yourself, don't add it. It's easy to add later. Let's not write code that is immediately unused.

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,681 @@
+        return sb.toString();
+    }
+
+    private final static String COMMENT_PREFIX = "##";
```
Actually I end up just removing this constant.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExonDB.java:
```diff
@@ -0,0 +1,370 @@
+     * right after the last exon in the range.
+     * </p>
+     */
+    class IndexRange {
```
you're right.  But there are many existing classes that do this, eg:
` org.apache.commons.lang3.Range`
`com.google.common.collect.Range`
`org.broadinstitute.hellbender.tools.picard.illumina.parse.Range`

use an existing one, don't write a new one

vruano replied:
Happy to reuse, yet you have other suggestions? The first two use Object generics. I am not in favor of change int(s) for Integer(s) ... I don't see this a premature optimization ... it is simply overkill IMO.

The *.illumina.parse.Range, a part from having domain specific things (e.g. throw IlluminaParserException) they are 1-based stuff when this one would be zero based.

Is then Ok if I create a new one user org.broadinstitute.hellbender.utils?

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonDB.java:
```diff
@@ -0,0 +1,361 @@
+
+    private int lastBinarySearchResult = -1;
+
+    private int cachedBinarySearch(final GenomeLoc location) {
```
can you add a comment about it? every reader of this code will ask the same question i did so a comment will help them understand why the code looks this way.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonDB.java:
```diff
@@ -0,0 +1,361 @@
+
+    @Override
+    public List<GenomeLoc> exons(final int from, final int to) {
+        if (from < 0)
```
premature optimization. Keep the human in mind, not the machine. 1 case is much easier to read than 3

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,681 @@
+            // Outputting progress into the logs:
+            if (index > lastProgressIndex && index % 1000 == 0)
+                logger.log(Level.INFO, String.format("Reached exon %d of %d at %s",
+                        index + 1, exonDB.exonCount(), exonDB.exon(lastProgressIndex = index)));
```
is this really intended an assignment to `lastProgressIndex` ? 
this is very frightening - hidden assignments like this are not good!

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExonDB.java:
```diff
@@ -0,0 +1,370 @@
+    /**
+     * Indicates that more than one exon was found when at most one was expected.
+     */
+    class AmbiguousExonException extends RuntimeException {
```
make this extend `GATKException` so users know where it's coming from

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ReadGroup.java:
```diff
@@ -0,0 +1,164 @@
+ *
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+class ReadGroup {
```
i'm not convinced. As it is now, it's just a fancy class for storing one String pointer. When you need more stuff, you'll add it. My guess, is YAGNI.

vruano replied:
Ok will supress.

akiezun commented on src/main/java/org/broadinstitute/hellbender/utils/Utils.java:
```diff
@@ -104,6 +104,31 @@ private static int getLastSpace(final CharSequence message, int width) {
+     * @throws NullPointerException if {@code separator} or {@code objects} is {@code null}.
+     * @return a string with the values separated by the separator
+     */
+    public static String join(final String separator, final Object ... objects) {
```
can you use `String.join` instead of this function?

vruano replied:
Actually String.join does not accept non CharSequence joinees so it is still kind of convenient to have this one around. I'm still using it in a couple of places so (avoiding ""+ x or String.valueOf(x)). I rather keep it since I have gone through the trouble and there are other similar operations in Utils.

akiezun commented on src/test/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCountIntegrationTest.java:
```diff
@@ -0,0 +1,514 @@
+ *
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+public class ExomeReadCountIntegrationTest extends CommandLineProgramTest {
```
Good tests but many methods are not well tested, for example:
calculateColumnName
resolveColumnNames
count

overall `ExomeReadCounts` has only 69% test coverage. Let's get it to 90%

vruano replied:
Coveralls reports over 91% on that file.

https://coveralls.io/builds/2276888/source?filename=src%2Fmain%2Fjava%2Forg%2Fbroadinstitute%2Fhellbender%2Ftools%2Fexome%2FExomeReadCounts.java

akiezun commented on src/test/java/org/broadinstitute/hellbender/tools/exome/ExonDBUnitTest.java:
```diff
@@ -0,0 +1,571 @@
+    }
+
+    private final static File REFERENCE_FILE = new File(getTestDataDir(),"test_reference.fasta");
+    @SuppressWarnings("unused")
```
if unused, why have it?

akiezun commented on src/test/java/org/broadinstitute/hellbender/tools/exome/ExonDBUnitTest.java:
```diff
@@ -0,0 +1,571 @@
+ *
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+public class ExonDBUnitTest {
```
overall coverage in IntervalBackedExonDB is only 43%. We need 90%

vruano replied:
I think you're are probably looking into an early commit. I'm pretty sure I achieved green (>90%) in all files.

akiezun commented on src/test/java/org/broadinstitute/hellbender/tools/exome/ReadGroupUnitTest.java:
```diff
@@ -0,0 +1,117 @@
+ *
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+public class ReadGroupUnitTest extends BaseTest {
```
incomplete test coverage

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ReadGroup.java:
```diff
@@ -0,0 +1,164 @@
+        this.sample = newSample;
+    }
+
+    private Sample sample;
```
aha! a mutable field. this make this class even more suspect. Mutable classes should not be treated as values (and the presence of hashcode and equals indicates a value.

akiezun commented on src/test/java/org/broadinstitute/hellbender/tools/exome/SampleBuilderUnitTest.java:
```diff
@@ -0,0 +1,222 @@
+ *
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+public class SampleBuilderUnitTest extends BaseTest {
```
incomplete test coverage for the builder

akiezun commented on src/test/java/org/broadinstitute/hellbender/tools/exome/SampleDBUnitTest.java:
```diff
@@ -0,0 +1,294 @@
+       return result.toArray(new Object[result.size()][]);
+    }
+
+//    @Test(dataProvider = "samFileHeaderWithDefaultSampleData", dependsOnMethods = {"testCreationWithDefaultSample","testSampleIds"})
```
delete dead code

akiezun commented:
review complete. Back to @vruano

vruano commented on src/main/java/org/broadinstitute/hellbender/cmdline/programgroups/ExomeAnalysisProgramGroup.java:
```diff
@@ -0,0 +1,66 @@
+ *
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+public class ExomeAnalysisProgramGroup implements CommandLineProgramGroup {
```
Ok.

vruano commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExonDB.java:
```diff
@@ -0,0 +1,370 @@
+     */
+    default void forEachExon(final GenomeLoc overlapRegion, final ExonTask<E> exonTask) {
+        if (exonTask == null)
+            throw new NullPointerException();
```
Ok

vruano commented:
The last commit addresses comments.

I have remove ReadGroups and Sample.Builder. I have not found a replacement for IndexRange and so I have kept it moving it to hellbender.utils so it can be reused by other. @akiezun If you insist in reusing another exiting class could you please point out to a more appropriate one (i.e. 0-based, without negative indexes and using primitive ints).

Do you want me to squash it with previous commits for re-review?

I also would like to change names of SampleDB and ExonDB to SampleCollection and ExonCollection but for now I keeping them so that you can compare changes better.

akiezun commented:
if there's no easy replacement for IndexRange let's keep it for now, but it's a kill candidate in the long run.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,881 @@
+    /**
+     * Short name for the {@link #cohortName} argument.
+     */
+    protected static final String COHORT_SHORT_NAME = "shortName";
```
hmm, "shortName" doesn't like a good name

akiezun replied:
i meant: "doesn't sound like a good name"

vruano replied:
Yeah accident.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,881 @@
+         *
+         * @return never {@code null}.
+         */
+        protected String friendlyName() {
```
should this just be `toString`?

vruano replied:
mmm... need to check that it does not interfere with parameter parsing.

vruano replied:
It does but I can see how to fix it.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,881 @@
+         * @param countColumnsFactory the count-columns manager factory to be used with
+         *                            this group-by option. Assumed not to be {@code null}.
+         */
+        GroupBy(final Function<ExomeReadCounts,? extends CountColumns> countColumnsFactory) {
```
why do this factory dance and not just instantiate the object?

vruano replied:
The enum is always static and thus I cannot make reference to the ExomeReadCounts.this

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,881 @@
+        final GenomeLoc readLocation = genomeLocFactory.createGenomeLoc(read);
+
+        final int columnIndex = countColumns.columnIndex(read);
+        if (columnIndex >= 0) { // < 0 would mean means that the read is to be ignored.
```
typo "mean means"

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,881 @@
+    /**
+     * Column totals
+     */
+    private long[] columnTotals;
```
this seems redundant with the `counts` matrix (just counts the sum of columns)

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,881 @@
+        final int exonCount = exonDB.exonCount();
+        final int columnCount = counts.length;
+        final int countColumnCount = countColumns.columnCount();
+        final int[] countBuffer = new int[countColumnCount];
```
this temp only lives inside the `i` loop. Move it inside to limit confusion of the reader.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,881 @@
+
+        logger.log(Level.INFO, "Writing counts ...");
+        final int exonCount = exonDB.exonCount();
+        final int columnCount = counts.length;
```
extracting this temp saves no time and adds code - inline it

vruano replied:
I think it makes the code more clear as 'columnCount' is more informative than 'counts.length'.

vruano replied:
Or perhaps you way saying about the line right below:

final int countColumnCount = countColumns.columnCount();

which indeed is totally redundant (countColumnCount == columnCount).

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,881 @@
+            return;
+        }
+
+        final long exonSize = exonDB.exomeSize();
```
typo `exonSize` should be `exomeSize`

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,881 @@
+     * Each element of this enum represent a different substitution pattern and
+     * provides operations to apply it to a given count column name string.
+     */
+    protected enum CountColumnNamePattern {
```
hmm this logic of three different ways of grouping is now duplicated to this enum (so we have 2 enums and 1 real class hierarchy). Can this all be collected in the class hierarchy?

vruano replied:
I see what you mean but not quite as there is the "Default" substitution pattern. Even if you would ge rid of that one there is no a 1-to-1 functional relationship between this enum and the CountColumns heriarchy as. 

For example, when grouping by READ_GROUP (thus ReadGroupCountColumns would get involve) you can use the sample id substitution pattern '{s}' to add the sample id to the head column name.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,881 @@
+        @Override
+        protected void checkToolState(final ExomeReadCounts tool) {
+            super.checkToolState(tool);
+            if (tool.groupBy != GroupBy.COHORT) {
```
yup, this check is a consequence of duplicating the grouping logic

vruano replied:
Is a paranoia check I will simply removed; it only can be trigger by a bug in the private/compilation-time resoved section of the class.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,881 @@
+    private static final class ReadGroupCountColumns extends CountColumns {
+
+        /**
+         * {@inheritDoc}
```
why does everyone else (2 other constructors in sibling classes) get a javadoc and this guy only gets a {@inheritDoc}?

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java:
```diff
@@ -0,0 +1,881 @@
+         * Returns the number of count columns headers.
+         * @return never {@code null}.
+         */
+        protected abstract int columnCount();
```
This method is tricky and may lead to problems. The only time it's called is from inside of `composeCountColumns` and so the natural implementation of `columnCount` will lead to a NPE (the natural implementation being `columnNames.size()`. How about making a field for it and store in in the constructor (you know the size at that time).

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExonDB.java:
```diff
@@ -0,0 +1,299 @@
+     * @throws IllegalArgumentException if {@code overlapRegion} is {@code null}.
+     * @throws AmbiguousExonException if the query location overlaps more than one exon.
+     */
+    default E exon(final GenomeLoc overlapRegion) {
```
I suggest moving away from GenomeLoc if only possible - we don't need a reference here and GenomeLoc is on the kill list. @droazen and @lbergelson can help on the alternatives.

vruano replied:
I talked to @droazen about this in advance an he was ok with keeping it for now until the time comes. I will do the work. 

I rather don't do it as part of this pull request.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExonDB.java:
```diff
@@ -0,0 +1,299 @@
+        for (int i = 0; i < exonCount; i++) {
+            result += location(i).size();
+        }
+        return result;
```
replace with a 1-liner like this `exons.stream().mapToIn(e -> e.size()).sum()`

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/ExonDB.java:
```diff
@@ -0,0 +1,299 @@
+     *
+     * @return never {@code null} but perhaps empty if there is no exon in this collection.
+     */
+    List<E> exons();
```
document whether this method returns the live list from the DB or a copy or an unmodifiable view. The API client needs to know this.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonDB.java:
```diff
@@ -0,0 +1,363 @@
+ *
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+public class IntervalBackedExonDB implements ExonDB<GenomeLoc> {
```
final

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonDB.java:
```diff
@@ -0,0 +1,363 @@
+    /**
+     * Sorted list of intervals.
+     */
+    private final List<GenomeLoc> sortedIntervals;
```
I wonder if `SortedSet` would not be a better type here to make clear that this is collection is sorted. Do you need random access to the data?

vruano replied:
Yeah I need it. I think that the name makes it clear enough; in fact that is what one can see when working somewhere else in the file; you need to click (cmd+B) to get to see the type.

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonDB.java:
```diff
@@ -0,0 +1,363 @@
+        if (intervals == null) {
+            throw new IllegalArgumentException("the input intervals cannot be null");
+        }
+        this.sortedIntervals = Collections.unmodifiableList(new ArrayList<>(intervals));
```
who enforces the sorting?

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonDB.java:
```diff
@@ -0,0 +1,363 @@
+    private int cachedBinarySearch(final GenomeLoc location) {
+
+        if (lastBinarySearchResult < 0) {
+            final int candidate = - lastBinarySearchResult - 1;
```
no space between `-` operator and the operand

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonDB.java:
```diff
@@ -0,0 +1,363 @@
+     * </p>
+     *
+     * <p>
+     *     This method uses the field {@link #lastBinarySearchResult} to cache the last result, thus is
```
thread safety is *not* a design consideration in Hellbender

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/Sample.java:
```diff
@@ -0,0 +1,97 @@
+    /**
+     * Read-groups by their id.
+     */
+    private final List<String> readGroups;
```
why not store as a set?

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/SampleDB.java:
```diff
@@ -0,0 +1,429 @@
+        final List<SAMReadGroupRecord> readGroupRecords = header.getReadGroups();
+        final int readGroupCount = readGroupRecords.size();
+
+        final Set<String> orphanReadGroups = new HashSet<>(readGroupCount);
```
what is this for?

vruano replied:
Supports for sample-id less read-groups. They can be assigned a default sample-id. Not currently using the functionality although is fully unit-tested. Do you want me to remove it?

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/SampleDB.java:
```diff
@@ -0,0 +1,429 @@
+     */
+    private Map<String, Set<String>> composeReadGroupsBySampleId(String defaultSampleId, boolean ignoreDefaultSampleWithoutOrphanReadGroups, List<SAMReadGroupRecord> readGroupRecords, int readGroupCount, Set<String> orphanReadGroups) {
+        final Map<String,Set<String>> readGroupsBySampleId = new HashMap<>(readGroupCount);
+        readGroupsBySampleId.put(null,orphanReadGroups );
```
this code looks very scary to me. what does it do and can we do it differently?

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/SampleDB.java:
```diff
@@ -0,0 +1,429 @@
+        if (readGroupId == null) {
+            throw new IllegalArgumentException("the input read-group id cannot be null");
+        }
+        final Integer result = readGroupIndexById.get(readGroupId);
```
use `readGroupIndexById.getOrDefault(readGroupId, -1)`

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/SampleDB.java:
```diff
@@ -0,0 +1,429 @@
+     * @return {@code null} if there is not such a sample.
+     */
+    protected Sample sample(final String id) {
+        final Integer index = sampleIndexById.get(id);
```
`Map.getOrDefault`

akiezun commented on src/main/java/org/broadinstitute/hellbender/tools/exome/SampleDB.java:
```diff
@@ -0,0 +1,429 @@
+                composeSampleIndexByGroupId(defaultSampleId, readGroupCount, orphanReadGroups,
+                        samples, sampleIndexById);
+
+        this.defaultSampleIndex = sampleIndexById.containsKey(defaultSampleId) ?
```
Map.getOrDefault

akiezun commented on src/main/java/org/broadinstitute/hellbender/utils/IndexRange.java:
```diff
@@ -0,0 +1,145 @@
+    public boolean equals(final Object other) {
+        if (other == this) {
+            return true;
+        } else if (other == null || !(other instanceof IndexRange)) {
```
`!(other instanceof IndexRange)` is enough - `null` always fails any `instanceof` check

akiezun commented on src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java:
```diff
@@ -769,4 +770,36 @@ private static ReferenceSequenceFile createReference(final File fastaFile) {
+            throw new IllegalArgumentException("the input locatable collection cannot be null");
+        };
+        final int size = locatables.size();
+        if (size == 0)
```
curly braces

akiezun commented on src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java:
```diff
@@ -769,4 +770,36 @@ private static ReferenceSequenceFile createReference(final File fastaFile) {
+            return Collections.emptyList();
+        final List<GenomeLoc> result = new ArrayList<>(locatables.size());
+        locatables.forEach(l -> {
+            if (l == null) {
```
can you extract this null check outside. Without it, the whole code turns into a nice 1-liner. And the null check can be done using `locatables.stream().anyMatch(l -> l == null)` which is also a nice 1liner

akiezun commented:
review done. back to @vruano for small fixes. The biggest issue is the fact that the groupBy logic is duplicated 3 times - 2 enums and a class hierarchy.

vruano commented:
Done with second round of reviews. 

@akiezun Differences are in  c6bfc3d.

After approved I would like to change SampleDB to SampleCollection and ExonDB to ExonCollection, please let me know if you object.

Also I have address a couple of bugs that I found on my way. Perhaps @lbergelson and or @droazen should review those. 2161c3a and 1504922

droazen commented:
Could you make those two bug fix commits into separate pull requests? Just create a branch for each one and then "git cherry-pick" each commit onto each branch.

vruano commented:
Ok, in progress.

vruano commented:
Ok will do those changes in a separate pull-request.

vruano commented:
In had remove those bug fixes from this one as per @droazen request. 
@akiezun  please re-review test past locally.

akiezun commented:
welcome to Hellbender!
